lib/active_model/serializer/adapter/json_api/deserialization.rb

```diff
@@ -0,0 +1,207 @@
+          #       associations: {
+          #         author: {
+          #           data: {
+          #             type: 'user',
+          #             id: 2
+          #           }
+          #         },
+          #         second_author: {
+          #           data: nil
+          #         },
+          #         comments: {
+          #           data: [{
+          #             type: 'comment',
+          #             id: 3
+          #           },{
+          #             type: 'comment',
+          #             id: 4
+          #           }]
+          #         }
+          #       }
+          #     }
+          #   }
+          #
+          #   parse(document) #=>
+          #     # {
+          #     #   title: 'Title 1',
+          #     #   date: '2015-12-20',
+          #     #   author_id: 2,
+          #     #   second_author_id: nil
+          #     #   comment_ids: [3, 4]
+          #     # }
+          #
+          #   parse(document, fields: [:title, :author, date: :published_at]) #=>
+          #     # {
+          #     #   title: 'Title 1',
+          #     #   published_at: '2015-12-20',
+          #     #   author_id: 2,
+          #     # }
+          #
+          # rubocop:disable Metrics/CyclomaticComplexity
+          def parse!(document, options = {})
+            fields = parse_fields_option(options[:fields])
+
+            document = document.dup.permit!.to_h if document.is_a?(ActionController::Parameters)
+
+            fail ArgumentError, 'Invalid payload' unless payload_valid?(document)
+
+            hash = {}
+
+            primary_data = document['data']
+
+            hash[:id] = primary_data['id'] if primary_data['id'] && (fields.nil? || fields[:id])
+
+            attributes = primary_data['attributes']
+            hash.merge!(parse_attributes(attributes, fields)) if attributes
+
+            relationships = primary_data['relationships']
+            hash.merge!(parse_relationships(relationships, fields)) if relationships
+
+            hash
+          end
+          # rubocop:enable Metrics/CyclomaticComplexity
+
+          # Safe version of parse!. Returns an empty hash instead of raising ArgumentError.
+          def parse(document, options = {})
+            parse!(document, options)
+          rescue ArgumentError
+            {}
+          end
+
+          # Checks whether a payload is compliant with the JSON API spec.
+          #
+          # @api private
+          # rubocop:disable Metrics/CyclomaticComplexity
+          def payload_valid?(payload)
+            return false unless payload.is_a?(Hash)
+            primary_data = payload['data']
+            return false unless primary_data.is_a?(Hash)
+            attributes = primary_data['attributes']
+            return false unless attributes.nil? || attributes.is_a?(Hash)
+            relationships = primary_data['relationships']
+            return false unless relationships.nil? || relationships.is_a?(Hash)
```

Comment:
Feel free to correct me here (it's a bit late and I've not fully got my head around the whole project) but I think this logic isn't what we need. On https://github.com/beauby/active_model_serializers/commit/6ba0a1d77ace7da1a86f2afe0a8b8389c3d89ef9 I seem have the two following test cases:

On a hash without relationships:
```ruby
(byebug) hash = {'data'=>{'attributes'=>{'name'=>'test'}, 'type'=>'events'}}                                        
{"data"=>{"attributes"=>{"name"=>"test"}, "type"=>"events"}}

(byebug) ActiveModel::Serializer::Adapter::JsonApi::Deserialization.parse(hash, fields:[:name])
*** NoMethodError Exception: undefined method `each' for nil:NilClass
```

On a hash with relationships:
```ruby
(byebug) hash = {'data'=>{'attributes'=>{'name'=>'test'}, 'type'=>'events', 'relationships' => {'author'=>{'data'=>nil}}}}
{"data"=>{"attributes"=>{"name"=>"test"}, "type"=>"events", "relationships"=>{"author"=>{"data"=>nil}}}}

(byebug) ActiveModel::Serializer::Adapter::JsonApi::Deserialization.parse(hash, fields:[:name])                     
{:name=>"test"}
```

Looking at the `unless` statements on line 106 and 108, if there is no relationship key in the hash as per the first test case, the `relationships` variable will be `nil` causing the `unless` statement to pass with no return of false.

It feels like the logic may need to be separated:
* If the relationships key does not exist then continue.
* If the relationships key exists, and it is a hash then iterate.
* If the relationships key exists, and it is not a hash then return false.